### ISSUE TITLE
How to update edges?

### ISSUE THREAD

**Issue by LeeHuan18** [pyg-team/pytorch_geometric#813]
Hi,
    Just like the nodes update ( and message passing ) by x_j = f(x_i), I want to also update the edges features e_j = f(node1, node2, e_j), with node1, node2 it's the two nodes edge e_j connect to. I don't want to use line_graph in my case. Is there anything like the MassagePassing  https://pytorch-geometric.readthedocs.io/en/latest/notes/create_gnn.html for the edges?

**Comment by fdiehl** [comment]
To the best of my knowledge, the best way of doing this is using the [`MetaLayer`](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.meta.MetaLayer) (depending on your use-case, simply ignore the `GlobalModel` part).

For example, for my recent work, I used something close to the example from `MetaLayer`:

```python
class EdgeModel(torch.nn.Module):
    def __init__(self, n_features, n_edge_features, hiddens, n_targets, residuals):
        super().__init__()
        self.residuals = residuals
        self.edge_mlp = nn.Sequential(
            nn.Linear(2 * n_features + n_edge_features, hiddens),
            nn.ReLU(),
            nn.Linear(hiddens, n_targets),
        )

    def forward(self, src, dest, edge_attr, u=None, batch=None):
        out = torch.cat([src, dest, edge_attr], 1)
        out = self.edge_mlp(out)
        if self.residuals:
            out = out + edge_attr
        return out


class NodeModel(torch.nn.Module):
    def __init__(self, n_features, n_edge_features, hiddens, n_targets, residuals):
        super(NodeModel, self).__init__()
        self.residuals = residuals
        self.node_mlp_1 = nn.Sequential(
            nn.Linear(n_features + n_edge_features, hiddens),
            nn.ReLU(),
            nn.Linear(hiddens, n_targets),
        )
        self.node_mlp_2 = nn.Sequential(
            nn.Linear(hiddens + n_features, hiddens),
            nn.ReLU(),
            nn.Linear(hiddens, n_targets),
        )

    def forward(self, x, edge_index, edge_attr, u, batch):
        row, col = edge_index
        out = torch.cat([x[col], edge_attr], dim=1)
        out = self.node_mlp_1(out)
        out = scatter_mean(out, row, dim=0, dim_size=x.size(0))
        out = torch.cat([x, out], dim=1)
        out = self.node_mlp_2(out)
        if self.residuals:
            out = out + x
        return out


def build_layer(self, n_hiddens, batchnorm, residuals):
    return geom_nn.MetaLayer(
        edge_model=EdgeModel(n_hiddens, n_hiddens, n_hiddens, n_hiddens, residuals=residuals),
        node_model=NodeModel(n_hiddens, n_hiddens, n_hiddens, n_hiddens, residuals=residuals),
    )

```

This can then be called using `x, edge_attr, _ = layer(x, edge_index, edge_attr=edge_attr)`.

Note that this didn't use or produce a global state, but did update the edge attributes.

**Comment by rusty1s** [comment]
Great answer! Following up on this, updating edge features is quite trivial without any helper functions, *e.g.*:
```python
row, col = edge_index
new_edge_attr = self.mlp(torch.cat([x[row], x[col], edge_attr], dim=-1))
```
Providing a more elegant API analog to `MessagePassing` is interesting though. Will think about it!

**Comment by tiberium24** [comment]
The problem for me is that there two edges for the same two nodes if the graph is undirected. That creates ambiguity. Which edge is the one I should pick? I only want the predicted value of one.

**Comment by Padarn** [comment]
@rusty1s Any further thoughts on this one? Would people be interested in a contribution if not? (I'm not an expert but can give it a go)

**Comment by rusty1s** [comment]
I personally like this idea. We could add an additional function, e.g., `edge_updater` to allow this directly in `MessagePassing`.

**Comment by Padarn** [comment]
Yeah agreed. Maybe can make a small POC to agree on an API first. Will try have time in the weekend.

**Comment by Padarn** [comment]
Well I didn't get around to this for a long time, but thought I'd come back to it.

After looking at the code a bit and trying to understand how it might be used I came to a few conclusions:

- Trying to add the functionality into 'propagate' would be a bit hard for two reasons 1) it would be hard to decide if you first do edge update or first do node update, or do them independently, 2) it would be hard to indicate to propagate which features were edge features as it takes `kwargs` and decodes it.
- Actually the solution provided by @rusty1s may be the best thing.

However I wrote a short example of how it could possible look if it was still desired as part of MessagePassing

https://github.com/pyg-team/pytorch_geometric/pull/3450/files

This is just a demonstration of how the API might look like. Thoughts?

**Comment by rusty1s** [comment]
Looks great, see https://github.com/pyg-team/pytorch_geometric/pull/3450#issuecomment-962911473. Let's move the discussion to the PR :)

**Comment by PabloVD** [comment]
Great answer @[fgerzer](https://github.com/fgerzer)! Is there a good reason to use residuals in that way, adding the input edge and node attributes to the result of the mlp? Can it act as a kind of skip connection?

> 
>     def forward(self, src, dest, edge_attr, u=None, batch=None):
>         out = torch.cat([src, dest, edge_attr], 1)
>         out = self.edge_mlp(out)
>         if self.residuals:
>             out = out + edge_attr
>         return out
>

**Comment by fgerzer** [comment]
This was indeed inspired by skip connections/jumping knowledge connections.

In my own experiments, JK connections on average improve the performance simple node classification graphs (+2.2pp on citeseer, +1.1pp on cora, +0.9pp on pubmed), but slightly decrease it for graph classification.

On more complex datasets, they are more worthwhile: On a traffic prediction dataset (with edge features, but not learning them), residuals reduced prediction error from 3.96m @ 5 seconds to 3.45m. For other experiments, I've just always included either a separate residual connection but have not conducted any deeper experiments.

**Comment by PabloVD** [comment]
Thank you for the detailed answer! I employed these residuals for node prediction, and the performance improved. I guess it is an advisable technique in general then.

**Comment by ArdaS2012** [comment]
I have a follow up question on the edge update. How can we easily aggregate edge features of one node with PyTorch (for example with mean or sum) ? Do you have some simple code snippet that explains this well? I didn’t quite understood it

**Comment by rusty1s** [comment]
```python
from torch_geometric.utils import scatter

out = scatter(edge_attr, edge_index[1], dim=0, dim_size=x.size(0))

**Comment by ArdaS2012** [comment]
> ```python
> from torch_geometric.utils import scatter
> 
> out = scatter(edge_attr, edge_index[1], dim=0, dim_size=x.size(0))
> ```

Thank you for answering! But then my new edge attributes have a different shape than before. Is it possible to keep the dimensions of edge attributes ?

**Comment by rusty1s** [comment]
Not sure what you mean. I thought you wanted to aggregate edge features across nodes? Here, `out` will be a tensor of shape `[num_nodes, num_edge_features]`.

**Comment by ArdaS2012** [comment]
I want to usw those updated/aggregated edge attributes in the model model as described above. And then of course x[row] or x[col] won’t be concatenated with edge attributes or „out“.

**Comment by rusty1s** [comment]
You can bring them back into edge-space via `out[edge_index[1]`. Does that work for you?

**Comment by ArdaS2012** [comment]
Yes this does the job for me thank you! Btw. I could simply use cat(x, out) instead of (x[col],out) if I don’t transform it back to edge space right ?

**Comment by fgerzer** [comment]
Yes. That'd give you a feature representation for each node that's "features of node concatenated with the aggregated features of all incoming edges."

**Comment by PabloVD** [comment]
Hi @egressybeni , 

Updating edge attributes in that way is one of the key elements in the Graph Network formalism, which is some sort of generalization of the standard message passing scheme. It is outlined [here](https://arxiv.org/abs/1806.01261), and based on the previous [Interaction Networks](https://arxiv.org/abs/1612.00222).

**Comment by egressybeni** [comment]
Thank you Pablo! That's very helpful :)
